File: lwjgl/src/main/java/com/darkraha/opengldemoj/gl/shaders/FragmentShaderBuilder.java
package com.darkraha.opengldemoj.gl.shaders;

public class FragmentShaderBuilder extends ShaderBuilderA {

    public FragmentShaderBuilder() {
        exOut = false;
    }

    public FragmentShaderBuilder exchangeData(boolean color, boolean texCoord, boolean lighting) {
        exColor = color;
        exTexCoord = texCoord;
        exLighting = lighting;
        return this;
    }

    @Override
    protected String buildDeclareOutCustom() {
        return "out vec4 fragColor;\n";
    }

    @Override
    protected String buildCalcCustom() {
        StringBuilder sb = new StringBuilder();

        if (exLighting) {
            sb.append("vec4 texelColor = texture(uSampler, exTexCoord);\n");
            sb.append("fragColor = vec4(texelColor.rgb * exLighting, texelColor.a);\n");
        } else if (exTexCoord) {
            sb.append("fragColor = texture(uSampler, exTexCoord);\n");
        } else if (exColor) {
            sb.append("fragColor = exColor;\n");
        }

        return sb.toString();
    }

    @Override
    protected String buildDeclareUniformCustom() {
        // todo multiple textures
        return "uniform sampler2D uSampler;\n";
    }
}

File: lwjgl/src/main/kotlin/com/darkraha/opengldemokt/gl/shaders/FragmentShaderBuilder.kt
package com.darkraha.opengldemokt.gl.shaders

import java.lang.StringBuilder

class FragmentShaderBuilder : ShaderBuilderA() {
    fun exchangeData(color: Boolean, texCoord: Boolean, lighting: Boolean): FragmentShaderBuilder {
        exColor = color
        exTexCoord = texCoord
        exLighting = lighting
        return this
    }

    override fun buildDeclareOutCustom(): String {
        return "out vec4 fragColor;\n"
    }

    override fun buildCalcCustom(): String {
        val sb = StringBuilder()
        if (exLighting) {
            sb.append("vec4 texelColor = texture(uSampler, exTexCoord);\n")
            sb.append("fragColor = vec4(texelColor.rgb * exLighting, texelColor.a);\n")
        } else if (exTexCoord) {
            sb.append("fragColor = texture(uSampler, exTexCoord);\n")
        } else if (exColor) {
            sb.append("fragColor = exColor;\n")
        }
        return sb.toString()
    }

    override fun buildDeclareUniformCustom(): String {
        // todo multiple textures
        return "uniform sampler2D uSampler;\n"
    }

    init {
        exOut = false
    }
}
